### PR TITLE
feat(snmp-discovery): Module/ModuleBay discovery via ENTITY-MIB (OBS-1828)

### PR DESCRIPTION
**Pull request by leoparente**
## Summary
- Vendor-neutral Module/ModuleBay discovery for SNMP backend via ENTITY-MIB, mirroring device-discovery PR #419 contract.
- Three-mode policy option: \`off\` (default — zero behaviour change), \`linecards\` (chassis-slot modules: linecards + supervisors), \`full\` (linecards + per-transceiver sub-bays + \`Interface.Module\` refs).
- VC-of-modular from day one — per-member dispatch via existing \`ChassisInventory\`; master keyed by lowest \`Members[0].ID\`.
- Sub-bay reconciler workaround: transceiver sub-bays emitted device-rooted (no \`Module: parentLinecard\` link). Per-port optic visibility preserved via \`Module.ModuleBay\`. Mirrors device-discovery's workaround pending Diode single-batch plan-and-apply.

## Approach
- Parallel pass in \`snmp-discovery/mapping/\` mirroring the \`chassis.go\` pattern. Three new files: \`module.go\` (types + classifier + extractor), \`module_routing.go\` (per-member dispatch + iface↔module map), \`module_translate.go\` (Diode entity emission).
- New \`ChassisModuleMapper\` no-op registers \`entPhysicalDescr\` + \`entPhysicalVendorType\` columns for the walk. Translation runs as a post-pass after \`TranslateAsStack\`.
- Classifier uses **both** containment-depth (under class=9 ancestor → transceiver candidate) **and** PID-prefix matching (SUP/SFP/QSFP/PSU/FAN). Both signals required for transceiver classification. PSU/Fan recognised but never emitted as module entities (label-only).
- \`Interface.Module\` ref enriched with bay matcher (Name + Position + chassis Device stub) so Diode reconciler resolves to existing standalone Module — mirrors device-discovery's Python stub shape exactly.
- New OTLP counters: \`modules_emitted{vendor, type}\`, \`module_bays_emitted{vendor}\`, \`modules_dropped{reason}\`.

## Validation
- 35 commits, 20 files, +2921 lines.
- \`go test ./...\` clean across 11 packages.
- \`go vet ./...\` clean. \`golangci-lint run\` 0 issues.
- TDD throughout: every feat task has failing-test-first → impl → run-to-pass commits.
- **E2E validated against orb-test-lab snmpsim** with a Cat-9404R-style \`.snmprec\` fixture (chassis + supervisor + linecard + transceiver + empty bay + PSU). All three modes work as designed:
  - **OFF:** 0 modules / 0 bays.
  - **LINECARDS:** 2 modules (Sup + Linecard), 2 bays, PSU filtered, transceiver filtered.
  - **FULL:** 3 modules (+ transceiver), 4 bays (+ empty Slot 3 + transceiver sub-bay), PSU filtered, \`Interface.Module\` populated to standalone transceiver Module via bay matcher.
- **Idempotency stress:** 5 sequential identical full-mode ingests produce byte-identical NetBox state (md5 stable), zero reconciler errors, IDs preserved, \`last_updated\` unchanged.

## Reviews completed before this PR
- 5 doc-review rounds (codex adversarial) on the design + plan before any code written — caught 22 issues, all patched in the docs (uncommitted) before implementation began.
- 2 post-implementation codex adversarial passes — caught 7 correctness issues (manufacturer source, non-recursive SubModules walk, PSU/Fan classifier suffix gap, port-container false-positive empty bay, missing string normalization, ambiguous Interface.Module stub, -FAN over-match). All patched.
- 1 Claude + 1 Codex normal review pass — caught 3 Medium correctness bugs (BayPosition source, chassis-rooted modules dropped, AliasMapFromOIDs nondeterminism) + 8 nits. All patched.
- 1 final codex pass on the Interface.Module enrichment + E2E double-ingest stress.

## Test plan
- [ ] Reviewer: read the spec context — the spec/plan docs live at \`docs/superpowers/{specs,plans}/2026-05-25-snmp-module-discovery-*.md\` (uncommitted by project rule; available on request).
- [ ] Reviewer: smoke test the \`mapping/module*\` unit suite locally.
- [ ] Reviewer: run E2E with a modular snmpsim fixture if possible (or rely on the recorded E2E artifacts).
- [ ] Reviewer: confirm device-discovery PR #419 contract parity for emitted Module/ModuleBay shape.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
@codex review

File: snmp-discovery/mapping/module_routing.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+			}
+			ifName, ok := ifIndexToName[ifIdx]
+			if !ok {
+				continue
+			}
+			mod, ok := emittedModules[e.EntIndex]
+			if !ok {
+				continue
+			}
+			out[ifName] = mod
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/mapping/module_routing.go]: Fixed in 7356c8f. Re-keyed `buildIfaceModuleMap` to use ifIndex (globally unique in the SNMP walk space) instead of ifName, eliminating the collision when VC members share canonical ifNames locally (Juniper VC, some Aruba stacks). The runner now looks up `ifaceModuleMap[strconv.Itoa(ifIndexByIface[iface])]` at attach time. Also dropped the now-unused `IfNameByIfIndex` helper and `ifIndexToName` parameter on `TranslateModulesWithAlias`. New regression test `TestBuildIfaceModuleMap_VCMembersWithSameIfNameDoNotCollide` pins the fix. E2E re-validated (3× idempotency stress, zero divergence).

**Comment by leoparente** [comment]
@codex review

File: snmp-discovery/mapping/module_routing.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+		ifIdx, err := strconv.Atoi(ifIdxStr)
+		if err != nil {
+			continue
+		}
+		candidates[entIdx] = append(candidates[entIdx], ifIdx)
+	}
+	out := make(map[string]string, len(candidates))
+	for entIdx, ifIdxs := range candidates {
+		sort.Ints(ifIdxs)
+		out[entIdx] = strconv.Itoa(ifIdxs[0])
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/mapping/module_routing.go]: Fixed in 4816c77. `AliasMapFromOIDs` now mirrors the chassis-routing precedence rule (`chassis_routing.go:188-220`): non-zero `entAliasLogicalIndexOrZero` rows win over the `.0` wildcard; within a tier, lowest logical-index wins; lowest ifIndex is the final tiebreaker. Two new tests pin the contract — `TestAliasMapFromOIDs_NonZeroLogicalIndexBeatsWildcard` (the actionable failing case: `.0` row at ifIndex 10101 vs logical=5 row at ifIndex 20201 — expects 20201 to win) and `TestAliasMapFromOIDs_LowestLogicalIndexAmongNonZero` (secondary-sort lock-in).

File: snmp-discovery/policy/runner.go [hunk 3]
```diff
@@ -544,6 +545,63 @@ func (r *Runner) queryTarget(ctx context.Context, target config.Target) ([]diode
...
+		// Attach Interface.Module on physical-port Interfaces for `full` mode.
+		// Order-safe because modules now precede interfaces in the slice.
+		// Key by ifIndex (globally unique in the SNMP walk space) — keying
+		// by Interface.Name would collide on VC members that reuse the same
+		// canonical name locally.
+		if len(ifaceModuleMap) > 0 {
+			for _, e := range entitiesForTarget {
+				iface, ok := e.(*diode.Interface)
+				if !ok {
+					continue
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/policy/runner.go]: Fixed in 5353cbe. Extracted the runner's inline attach loop into `mapping.AttachIfaceModules(entities, ifaceModuleMap, ifIndexByIface)` which walks three referrer shapes: top-level `*diode.Interface`, `*diode.IPAddress.AssignedObject`, and `*diode.MACAddress.AssignedObject`. Verified `ifIndexByIface` already covers nested-only interfaces via the IPAddress mapper's `GetOrCreateEntity(InterfaceEntityType, ifIndex)` call at `mappers.go:308` (no second source of truth needed). Five new tests cover both nested shapes, the top-level regression guard, the graceful-miss case, and the empty-maps no-op. E2E re-validated against the lab fixture — interface 186 (TenGigabitEthernet2/0/1) still carries `module: 88`.

**Comment by leoparente** [comment]
@codex review

File: snmp-discovery/policy/mapping.yaml [hunk 4]
```diff
@@ -126,6 +126,12 @@ entries:
     identifier_size: 2
     description: "ENTITY-MIB entPhysicalTable (chassis physical inventory for stack detection)"
     mapping_entries:
+      - oid: ".1.3.6.1.2.1.47.1.1.1.1.2"   # entPhysicalDescr
+        entity: "chassis_module"
+        field: "descr"
+      - oid: ".1.3.6.1.2.1.47.1.1.1.1.3"   # entPhysicalVendorType
+        entity: "chassis_module"
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/policy/mapping.yaml:133]: Fixed in d0f5d59. Gated chassis_module entry expansion inside `objectIDsForVendor` (mapping.go:1502): when `config.Options.ModuleDiscoveryMode() == DiscoverModulesOff`, child entries with `entity: chassis_module` are skipped before they reach the walker. The two new columns (`.1.3.6.1.2.1.47.1.1.1.1.2` and `.1.3.6.1.2.1.47.1.1.1.1.3`) drop out of the OID set entirely. `linecards` and `full` modes pass through unchanged.

Test `TestGenericObjectIDs_ChassisModuleColumnsGatedByDiscoverModules` has four sub-cases (unset / off / linecards / full) asserting the gating in both directions. E2E: off-mode walk dropped from 37 → 35 OIDs and a grep over the run log confirms no `object_id=.1.3.6.1.2.1.47.1.1.1.1.[23].x` PDUs reach the mapper. Full-mode walk count stays at 37 and modules still emit correctly on the lab fixture (`module: 88` on TenGigabitEthernet2/0/1).

**Comment by leoparente** [comment]
@codex review

**Comment by leoparente** [comment]
@codex review

File: snmp-discovery/mapping/module_translate.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+// emitModule constructs a Module entity attached to its ModuleBay.
+// Carries Device (NetBox matching scope) and a ModuleType built from
+// the PID (Model) + the manufacturer resolved from the emitted Device.
+// Manufacturer precedence: Device.DeviceType.Manufacturer.Name first
+// (so the ModuleType label always matches what NetBox sees on the
+// owning device), then the policy-level defaults, finally "Unknown".
+// Sharing vendorFromDevice with the metrics path keeps the label and
+// the emitted entity identical strings.
+func emitModule(device *diode.Device, bay *diode.ModuleBay, m ModuleEntry, defaults *config.Defaults) *diode.Module {
+	model := modelOrUnknown(m.Model)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/mapping/module_translate.go]: Fixed in 6259439. New helper `modelOrVendorType(m.Model, m.VendorType)` in `module_translate.go` mirrors `classifyModule`'s existing fallback chain: `strings.TrimSpace(model) → strings.TrimSpace(vendorType) → "Unknown"`. `emitModule` now uses it instead of the old `modelOrUnknown` (removed; had no other callers). Aruba CX–shape modules (blank `entPhysicalModelName`, populated `entPhysicalVendorType`) now emit with the real PID in `ModuleType.Model` rather than `"Unknown"`. New test `TestTranslateModules_ModuleTypeModel_FallsBackToVendorTypeWhenModelBlank`.

File: snmp-discovery/mapping/module.go [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+
+	// Process class=9 rows in EntIndex-ascending order so dedup
+	// "first occurrence wins" is deterministic.
+	classNineIdxs := make([]string, 0, len(byIdx))
+	for _, r := range byIdx {
+		if r.Class == entPhysicalClassModule {
+			classNineIdxs = append(classNineIdxs, r.EntIndex)
+		}
+	}
+	sort.Strings(classNineIdxs)
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/mapping/module.go]: Fixed in 803c0bf. Replaced `sort.Strings(classNineIdxs)` with a `sort.Slice` over `strconv.Atoi` that falls back to lex compare for non-numeric edge cases. ENTITY-MIB indexes are numeric, so lex order put `"10"` before `"9"` and the dedup winner was wrong — now EntIndex 9 correctly wins over EntIndex 10 on duplicate-serial collisions. New test `TestExtractModuleInventory_DedupUsesNumericEntIndexOrder`.

**Review by jajeffries — APPROVED** [review]

File: snmp-discovery/policy/runner.go [hunk 7]
```diff
@@ -544,6 +544,53 @@ func (r *Runner) queryTarget(ctx context.Context, target config.Target) ([]diode
...
+		// Order-safe because modules now precede interfaces in the slice.
+		// Key by ifIndex (globally unique in the SNMP walk space) — keying
+		// by Interface.Name would collide on VC members that reuse the same
+		// canonical name locally. AttachIfaceModules also walks
+		// IPAddress.AssignedObject / MACAddress.AssignedObject so L3
+		// routed-port interfaces (filtered out of the top-level entity
+		// slice by MapObjectIDsToEntity.getAssignedInterfaces) still get
+		// their module set.
+		mapping.AttachIfaceModules(entitiesForTarget, ifaceModuleMap, ifIndexByIface)
+	}
```
[the remark this answers is not in the document]
Comment (leoparente) [reply snmp-discovery/policy/runner.go]: Fixed in 3401af1. Replaced the bare `{ ... }` block with `if r.config.Options.ModuleDiscoveryMode() != config.DiscoverModulesOff { ... }` so the entire module-emission stage — including `ChassisInventoryFromOIDs`, `MemberDevicesFromEntities`, `AliasMapFromOIDs`, `SpliceModulesAfterDevices`, and `AttachIfaceModules` — is skipped on the default path. `off` mode is now a true no-op (zero map scans, zero allocations from this block).
